elasticdl/pkg/ps/optimizer.go

```diff
@@ -130,11 +139,106 @@ func (opt *AdamOptimizer) InitNonEmbeddingParam(name string, dim []int64) {
 	opt.maxSquare.NonEmbeddingParam[name] = &common.Tensor{name, make([]float32, size, size), dim, nil}
 }
 
-// NewOptimizer creates an optimizer instance
-func NewOptimizer(opt string, lr float32) Optimizer {
-	// TODO(qijun) only support SGD now
-	if opt == "SGD" {
-		return NewSGDOptimizer(lr)
+const (
+	optTypeSGD     = "SGD"
+	optTypeAdam    = "Adam"
+	optArgLR       = "learning_rate"
+	optArgMomentum = "momentum"
+	optArgNesterov = "nesterov"
+	optArgBeta1    = "beta_1"
+	optArgBeta2    = "beta_2"
+	optArgEpsilon  = "epsilon"
+	optArgAmsgrad  = "amsgrad"
+)
+
+var optArgumentsMap = map[string][]string{
+	"SGD":  []string{optArgLR, optArgMomentum, optArgNesterov},
+	"Adam": []string{optArgLR, optArgBeta1, optArgBeta2, optArgEpsilon, optArgAmsgrad},
+}
+
+// parseOptArgs parses optimizer arguments according to optimizer type
+func parseOptArgs(optType string, optArgs string) (map[string]string, error) {
+	// parse arguments to map
+	argsMap := make(map[string]string)
+	for _, args := range strings.Split(optArgs, ";") {
+		if args == "" {
+			continue
+		} else {
+			arr := strings.Split(args, "=")
+			argsMap[arr[0]] = arr[1]
+		}
+	}
+
+	// check argument names
+	for _, argName := range optArgumentsMap[optType] {
+		if _, ok := argsMap[argName]; !ok {
+			return nil, fmt.Errorf("Args passed to ps should contain %s", argName)
+		}
+	}
+	if len(argsMap) != len(optArgumentsMap[optType]) {
+		return nil, fmt.Errorf("Args passed to ps contain redundant items: %v", argsMap)
```

Comment:
The other case is that args passed to ps do not have all the required items. May change to:
"Args do not match %v optimizer", optType

Reply:
Above code checked that args passed to ps have all the required items.

Reply:
I see.